### PR TITLE
fix: always pass a string as a value to ace editor

### LINKED ISSUE #13540: [SQLlab]Parameters Modal Not Appearing
#### How to reproduce the bug

Prereq: set ENABLE_TEMPLATE_PROCESSING = True 
1. Go to SQL Lab 
2. Click the overflow menu on the right of the toolbar 
3. Click "Parameters" 

### Expected results
Modal appears to set Parameters 

### Actual results
No modal appears and screen is blank other than the toolbar 

#### Screenshots

image

### Environment
- superset version: `master`

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by eschutho**
### SUMMARY
Fixes issue with template params modal not opening when value was null. 
I converted one file to tsx so that we can put a null check on the code argument in TemplateParamsEditor, but it won't help much until the rest of the files are converted. I also made the initial state undefined instead of null so that we can use default args downstream. Lastly, I put in a check at the ace editor itself to never accept a null or undefined value, but pass an empty string in that case. 

fixes https://github.com/apache/superset/issues/13540

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
see #13540 for before state (no modal opens)

after:
Screenshot_3_10_21__7_32_PM

### TEST PLAN
unit tests aren't completely reliable when it comes to passing a null value to ace editor. It was difficult to reproduce the issue in the tests. I wrote tests around the new code to ensure that we are using undefined instead of null.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by junlincc** [comment]
Tested locally and learned something new today, LGTM! Thanks for the fix! 🙏 @eschutho 
With FF enabled, I can type and delete in both sql editor and para modal ✅
i can open a new tab with ctrl+t ✅
I am able to type, close, reopen the modal and the data persists ✅

Functionality ✅
https://user-images.githubusercontent.com/67837651/110876581-a201ee80-828c-11eb-9b0d-ecfc5221044d.mov

**Comment by eschutho** [comment]
flaky test?

**Comment by junlincc** [comment]
> flaky test?

i think so, we had to disable it in our PRs.

File: superset-frontend/src/components/AsyncAceEditor/index.tsx [hunk 1]
```diff
@@ -150,7 +151,8 @@ export default function AsyncAceEditor(
             mode={mode}
             theme={theme}
             tabSize={tabSize}
-            {...props}
+            defaultValue={defaultValue}
+            {...{ ...props, value: props.value || '' }}
```
Comment (betodealmeida) [inline superset-frontend/src/components/AsyncAceEditor/index.tsx:155]: The syntax here looks alien. 😄 

Why not:

```suggestion
            {...props}
            value={props.value || '' }
```

File: superset-frontend/src/SqlLab/components/TemplateParamsEditor.tsx [hunk 2]
```diff
@@ -26,34 +25,32 @@ import { debounce } from 'lodash';
...
-function TemplateParamsEditor({ code, language, onChange }) {
-  const [parsedJSON, setParsedJSON] = useState();
+function TemplateParamsEditor({
+  code = '{}',
+  language,
+  onChange = () => {},
+}: {
+  code: string;
+  language: 'yaml' | 'json';
+  onChange: (...args: any) => any;
```
Comment (zhaoyongjie) [inline superset-frontend/src/SqlLab/components/TemplateParamsEditor.tsx:43]: how about this?
```suggestion
  onChange: () => void;
```

Comment (zhaoyongjie) [reply superset-frontend/src/SqlLab/components/TemplateParamsEditor.tsx:43]: When we are convert `SqlEditor.jsx` to the typescript, we can add function parameter type again. 

https://github.com/apache/superset/blob/d439da2fe00c0a16edfec3fb5354d8e8f4da470b/superset-frontend/src/SqlLab/components/SqlEditor.jsx#L512-L518

Comment (eschutho) [reply superset-frontend/src/SqlLab/components/TemplateParamsEditor.tsx]: We should be getting to it soon, and then, yes we'll be able to manage types of props better, ya!

File: superset-frontend/src/components/AsyncAceEditor/index.tsx (hunk 1 above)
Comment (eschutho) [reply superset-frontend/src/components/AsyncAceEditor/index.tsx]: ha, yeah, that was my first implementation, but often people like to see the "rest" props at the end, and so I anticipated someone changing this order in the context of "cleanup". I think I found a way around this.. lmk what you think.

**Review by zhaoyongjie — APPROVED** [review]
LGTM

**Comment by eschutho** [comment]
🏷 ready-to-merge

**Review by pkdotson — APPROVED** [review]

**Pull request merged**